test/extended/kubelet/userns.go

```diff
@@ -0,0 +1,361 @@
+	)
+	g.Context("[apigroup:config.openshift.io]", func() {
+		g.BeforeEach(func() {
+			exutil.PreTestDump()
+			//TODO remove when user namespaces go GA
+			if !exutil.IsTechPreviewNoUpgrade(oc) {
+				g.Skip("the test is not expected to work within Tech Preview disabled clusters")
+			}
+		})
+
+		g.AfterEach(func() {
+			if g.CurrentSpecReport().Failed() {
+				exutil.DumpPodStates(oc)
+				exutil.DumpConfigMapStates(oc)
+				exutil.DumpPodLogsStartingWith("", oc)
+			}
+		})
+		// This test verifies the kubelet does not create a user namespace pod if it doesn't support them.
+		// There are two cases that are being protected against here:
+		// - If the kubelet doesn't support user namespaces because of version skew
+		// - If the kubelet doesn't support user namespaces because of a FeatureGate rollout/upgrade
+		// In either of these cases, the root cause that needs to be verified is the kubelet doesn't have the feature gate
+		// on, but the KAS does.
+		// Since we can't easily test this through the openshift API, we (hackily) use a MachineConfig to create a drop-in
+		// kubelet service configuration. This configuration will add the `--feature-gate="UserNamespacesSupport=false"` flag
+		// to the kubelet's CLI, thus disabling the feature gate.
+		// Then, we check that the kubelet fails to create a pod that requests a user namespace.
+		// A lot of the code is adapted from MCO e2e tests, as well as an e2e-node test case that verifies the same.
+		g.It("Kubelet should fail to create a user namespaced pod if featuregate is off [Serial]", func() {
+			// instead of a bunch of individual defers, we can run through all of them in a single one
+			const mcpName = "userns"
+			cleanupFuncs := make([]func(), 0)
+			defer func() {
+				for _, f := range cleanupFuncs {
+					f()
+				}
+			}()
+			originalMCName := getMCName(oc, workerMCPName)
+			g.By(fmt.Sprintf("Starting with MC name %s", originalMCName))
+
+			// create mcp
+			cleanupFuncs = append(cleanupFuncs, createMCP(oc, mcpName))
+
+			// create mc and wait for it to roll out
+			mcName, deleteMCFunc := createKubeletDropinMC(oc, originalMCName, mcpName)
+			cleanupFuncs = append(cleanupFuncs, deleteMCFunc)
+
+			// label random node in worker pool
+			cleanupFuncs = append(cleanupFuncs, labelRandomNodeFromPool(oc, mcpName, workerMCPName, mcName, originalMCName))
+
+			// create pod
+			pod := podframework.MakePod(oc.Namespace(), map[string]string{mcpNameToRole(mcpName): ""}, nil, admissionapi.LevelBaseline, "")
+			falseVar := false
+			pod.Spec.HostUsers = &falseVar
+			g.By("Creating user namespaced pod")
+			pod, err := oc.KubeClient().CoreV1().Pods(oc.Namespace()).Create(context.Background(), pod, metav1.CreateOptions{})
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			// Pod should stay in pending
+			// Events would be a better way to tell this, as we could actually read the event from the kubelet saying the exact issue,
+			// but events are best-effort and can't be relied on.
```

Comment:
as a practical matter, they tend to be present in our CI runs.  Perhaps if the event is present end the test early?